### PR TITLE
[XPTI] 64-bit universal ID rework to use a tuple of three 64-bit components

### PR DESCRIPTION
**Pull request by tovinkere**
With the requirement of supporting 64-bit hash generation from code location+line number or kernel name + back trace + address or kernel name + address, the universal ID manages three 64-bit values representing each of these scenarios to generate a hash. Updated the unit tests and performance tests to use the new data structure and hash generation algorithm, including a simulation of 10000-1000000 million trace points that result in no collision.

### CODE REVIEW

File: xpti/include/xpti_data_types.h [hunk 1]
```diff
@@ -56,35 +133,38 @@ using metadata_t = std::unordered_map<string_id_t, string_id_t>;
...
-/// Unique_id MUST be propagated downstream to the OpenCL runtime to ensure the
-/// associations back to the sources. This requires elp from the compiler
+/// Unique_id MUST be propagated downstream to the OpenCL runtime to ensure
+/// the associations back to the sources. This requires elp from the compiler
 /// front-end.
 ///
 struct payload_t {
   /// String ID of the function/kernel name; only valid if 'name' is valid
-  string_id_t name_sid = invalid_id;
+  // string_id_t name_sid = invalid_id;
```
Comment (alexbatashev) [inline xpti/include/xpti_data_types.h:154]: Should we just remove these four lines?

File: xpti/include/xpti_data_types.h [hunk 2]
```diff
@@ -102,53 +184,67 @@ struct payload_t {
...
   payload_t(void *codeptr) {
     code_ptr_va = codeptr;
-    name_sid = invalid_id;        ///< Invalid string ID
-    source_file_sid = invalid_id; ///< Invalid string ID
-    name = nullptr;               ///< Invalid name string pointer
-    source_file = nullptr;        ///< Invalid source file string pointer
-    line_no = invalid_id;         ///< Invalid line number
-    column_no = invalid_id;       ///< Invalid column number
+    // name_sid = invalid_id;        ///< Invalid string ID
+    // source_file_sid = invalid_id; ///< Invalid string ID
```
Comment (alexbatashev) [inline xpti/include/xpti_data_types.h:188]:
```suggestion
```

File: xpti/include/xpti_data_types.h [hunk 3]
```diff
@@ -102,53 +184,67 @@ struct payload_t {
...
   payload_t(const char *func_name) {
     code_ptr_va = nullptr;
-    name_sid = invalid_id;        ///< Invalid string ID
-    source_file_sid = invalid_id; ///< Invalid string ID
-    name = func_name;             ///< Invalid name string pointer
-    source_file = nullptr;        ///< Invalid source file string pointer
-    line_no = invalid_id;         ///< Invalid line number
-    column_no = invalid_id;       ///< Invalid column number
+    // name_sid = invalid_id;        ///< Invalid string ID
+    // source_file_sid = invalid_id; ///< Invalid string ID
```
Comment (alexbatashev) [inline xpti/include/xpti_data_types.h:204]:
```suggestion
```

File: xpti/include/xpti_data_types.h [hunk 4]
```diff
@@ -102,53 +184,67 @@ struct payload_t {
...
   payload_t(const char *func_name, void *codeptr) {
     code_ptr_va = codeptr;
-    name_sid = invalid_id;        ///< Invalid string ID
-    source_file_sid = invalid_id; ///< Invalid string ID
-    name = func_name;             ///< Invalid name string pointer
-    source_file = nullptr;        ///< Invalid source file string pointer
-    line_no = invalid_id;         ///< Invalid line number
-    column_no = invalid_id;       ///< Invalid column number
+    // name_sid = invalid_id;        ///< Invalid string ID
+    // source_file_sid = invalid_id; ///< Invalid string ID
```
Comment (alexbatashev) [inline xpti/include/xpti_data_types.h:213]:
```suggestion
```

File: xptifw/CMakeLists.txt [hunk 5]
```diff
@@ -10,7 +10,7 @@ set(XPTIFW_DIR ${CMAKE_CURRENT_LIST_DIR})
 set(XPTI_DIR ${CMAKE_CURRENT_LIST_DIR}/../xpti)
 
 # Create a soft option for enabling the use of TBB
-option(XPTI_ENABLE_TBB "Enable TBB in the framework" OFF)
+option(XPTI_ENABLE_TBB "Enable TBB in the framework" ON)
```
Comment (alexbatashev) [inline xptifw/CMakeLists.txt:13]:
```suggestion
option(XPTI_ENABLE_TBB "Enable TBB in the framework" OFF)
```

File: xptifw/src/xpti_trace_framework.cpp [hunk 6]
```diff
@@ -532,107 +446,61 @@ class Tracepoints {
...
+      Event->data_id = Event->source_id = Event->target_id = 0;
+      Event->instance_id = 1;
+      Event->global_user_data = nullptr;
+      Event->event_type = (uint16_t)xpti::trace_event_type_t::unknown_event;
+      Event->activity_type =
+          (uint16_t)xpti::trace_activity_type_t::unknown_activity;
+      *InstanceNo = Event->instance_id;
+      return Event;
     }
-    return nullptr;
```
Comment (alexbatashev) [inline xptifw/src/xpti_trace_framework.cpp:617]: It looks like all cases are handled, but I'm still not sure if all compilers can process this code without firing a warning about non-void function without return statement.

**Review by alexbatashev — COMMENTED** [review]
Amazing work!

File: xpti/include/xpti_data_types.h (hunk 1 above)
Comment (tovinkere) [reply xpti/include/xpti_data_types.h]: Yes, they have been removed.

File: xptifw/src/xpti_trace_framework.cpp (hunk 6 above)
Comment (tovinkere) [reply xptifw/src/xpti_trace_framework.cpp:617]: If/else return values in all cases and the compilers did not complain. I prefer this as any changes in the future that accidentally delete a return line will cause the compiler to complain.

File: xpti/include/xpti_data_types.h [hunk 7]
```diff
@@ -11,24 +11,101 @@
 #include <string>
```
[the remark this answers is not in the document]
Comment (tovinkere) [reply xpti/include/xpti_data_types.h:12]: Good catch - I am surprised I didn't get a complaint on Linux. Done!

[the remark this answers is not in the document]
Comment (alexbatashev) [reply xpti/include/xpti_data_types.h:12]: Probably some other header includes it

**Review by alexbatashev — DISMISSED** [review]

**Review by alexbatashev — APPROVED** [review]

File: xptifw/src/xpti_trace_framework.cpp [hunk 8]
```diff
@@ -412,107 +397,36 @@ class Tracepoints {
...
+          MStringTableRef.add(Payload->stack_trace, &Payload->stack_trace);
     }
+    // Pack the 1st 64-bit value with string ID from source file name and line
+    // number; pack the 2nd 54-bit value with stack backtrace string ID and the
+    // kernel name string ID
+    Payload->uid.p1 = XPTI_PACK32_RET64(source_id, line_no);
+    Payload->uid.p2 = XPTI_PACK32_RET64(stack_id, name_id);
+    // The code pointer for the kernel is already in 64-bit format
+    if ((Payload->flags &
+         static_cast<uint64_t>(payload_flag_t::CodePointerAvailable)))
```
Comment (andykaylor) [inline xptifw/src/xpti_trace_framework.cpp:422]: It looks like you have an extra set of parentheses here.

**Review by andykaylor — APPROVED** [review]
I just have a few minor comments. You can ignore them if you don't have anything else to change.

File: xptifw/CMakeLists.txt [hunk 9]
```diff
@@ -52,6 +52,6 @@ endif()
 
 # The tests in basic_test are written using TBB, so these tests are enabled
 # only if TBB has been enabled.
-if (XPTI_ENABLE_TBB)
+if (0)
```
[the remark this answers is not in the document]
Comment (tovinkere) [reply xptifw/CMakeLists.txt:55]: Currently, TBB is required for exercising multi-threaded tests and marginal performance improvement of some data structure access in multi-threaded mode. However, XPTI does not require TBB to be fully functional. The reason for disabling this feature is that the Windows builds have a failure while linking the multi-threaded tests and once we resolve them, we should be able to turn them on.

File: xptifw/src/xpti_trace_framework.cpp (hunk 8 above)
Comment (tovinkere) [reply xptifw/src/xpti_trace_framework.cpp:422]: Thanks! Will address in the next PR as it is cosmetic.

**Pull request merged**